groups/bsl/bslstl/bslstl_optional.h

```diff
@@ -0,0 +1,1571 @@
+template <class TYPE>
+template <class... ARGS>
+inline
+void Optional_DataImp<TYPE>::emplace(
+                         bslma::Allocator                           *allocator,
+                         BSLS_COMPILERFEATURES_FORWARD_REF(ARGS)...  args)
+{
+    reset();
+    BloombergLP::bslma::ConstructionUtil::construct(
+        d_buffer.address(),
+        allocator,
+        BSLS_COMPILERFEATURES_FORWARD(ARGS, args)...);
+    d_hasValue = true;
+}
+
+#if defined(BSLS_COMPILERFEATURES_SUPPORT_GENERALIZED_INITIALIZERS)
+template <class TYPE>
+template <class INIT_LIST_TYPE, class... ARGS>
+void Optional_DataImp<TYPE>::emplace(
+                         bslma::Allocator                           *allocator,
+                         std::initializer_list<INIT_LIST_TYPE>       il,
+                         BSLS_COMPILERFEATURES_FORWARD_REF(ARGS)...  args)
+{
+    reset();
+    BloombergLP::bslma::ConstructionUtil::construct(
+        d_buffer.address(),
+        allocator,
+        il,
+        BSLS_COMPILERFEATURES_FORWARD(ARGS, args)...);
+    d_hasValue = true;
+}
+#endif
+// }}} END GENERATED CODE
+#endif
+
+template <class TYPE>
+void Optional_DataImp<TYPE>::reset() BSLS_KEYWORD_NOEXCEPT
+{
+    if (d_hasValue) {
+        bslma::DestructionUtil::destroy(&(d_buffer.object()));
+    }
+    d_hasValue = false;
+}
+
+#if defined(BSLS_COMPILERFEATURES_SUPPORT_REF_QUALIFIERS)
+template <class TYPE>
+inline
+TYPE& Optional_DataImp<TYPE>::value() &
+{
+    BSLS_ASSERT(d_hasValue);
+
+    return d_buffer.object();
+}
+
+template <class TYPE>
+inline
+TYPE&& Optional_DataImp<TYPE>::value() &&
+{
+    BSLS_ASSERT(d_hasValue);
+
+    return std::move(d_buffer.object());
+}
+#else
+template <class TYPE>
+inline
+TYPE& Optional_DataImp<TYPE>::value()
+{
+    BSLS_ASSERT(d_hasValue);
+
+    return d_buffer.object();
+}
+#endif  //defined(BSLS_COMPILERFEATURES_SUPPORT_REF_QUALIFIERS)
+
+//ACCESSORS
+template <class TYPE>
+inline
+bool Optional_DataImp<TYPE>::hasValue() const BSLS_KEYWORD_NOEXCEPT
+{
+    return d_hasValue;
+}
+
+#if defined(BSLS_COMPILERFEATURES_SUPPORT_REF_QUALIFIERS)
+template <class TYPE>
+inline
+const TYPE& Optional_DataImp<TYPE>::value() const &
+{
+    BSLS_ASSERT(d_hasValue);
+
+    return d_buffer.object();
+}
+
+template <class TYPE>
+inline
+const TYPE&& Optional_DataImp<TYPE>::value() const &&
+{
+    BSLS_ASSERT(d_hasValue);
+
+    return std::move(d_buffer.object());
+}
+#else
+template <class TYPE>
+inline
+const TYPE& Optional_DataImp<TYPE>::value() const
+{
+    BSLS_ASSERT(d_hasValue);
+
+    return d_buffer.object();
+}
+#endif  //defined(BSLS_COMPILERFEATURES_SUPPORT_REF_QUALIFIERS)
+
+                            // ===================
+                            // class Optional_Data
+                            // ===================
+
+// CREATORS
+template <class TYPE, bool IS_TRIVIALLY_DESTRUCTIBLE>
+Optional_Data<TYPE, IS_TRIVIALLY_DESTRUCTIBLE>::~Optional_Data()
+{
+    this->reset();
```

Comment:
Just for consistency with other calls to member functions
```suggestion
    reset();
```

Reply:
"this->" must be present - reset is a member of a type dependant base class.